File: infra/gateway/internal/feed/orderbook.go
package feed

import (
	"fmt"
	"math"
	"sort"
	"sync"
	"time"
)

// PriceLevel represents a single bid or ask level in an order book.
type PriceLevel struct {
	Price float64
	Size  float64
}

// OrderBook is an in-memory bid/ask order book for a single symbol.
// It maintains bids sorted descending and asks sorted ascending.
type OrderBook struct {
	mu     sync.RWMutex
	Symbol string
	Bids   []PriceLevel // descending by price
	Asks   []PriceLevel // ascending by price
	SeqNum int64
	AsOf   time.Time
}

// NewOrderBook creates an empty OrderBook.
func NewOrderBook(symbol string) *OrderBook {
	return &OrderBook{Symbol: symbol}
}

// UpdateBid sets the size at a bid price level (removes if size == 0).
func (ob *OrderBook) UpdateBid(price, size float64) {
	ob.mu.Lock()
	defer ob.mu.Unlock()
	ob.Bids = updateLevel(ob.Bids, price, size, false)
	ob.AsOf = time.Now()
}

// UpdateAsk sets the size at an ask price level (removes if size == 0).
func (ob *OrderBook) UpdateAsk(price, size float64) {
	ob.mu.Lock()
	defer ob.mu.Unlock()
	ob.Asks = updateLevel(ob.Asks, price, size, true)
	ob.AsOf = time.Now()
}

// BestBid returns the highest bid price and size, or (0,0) if empty.
func (ob *OrderBook) BestBid() (price, size float64) {
	ob.mu.RLock()
	defer ob.mu.RUnlock()
	if len(ob.Bids) == 0 {
		return 0, 0
	}
	return ob.Bids[0].Price, ob.Bids[0].Size
}

// BestAsk returns the lowest ask price and size, or (0,0) if empty.
func (ob *OrderBook) BestAsk() (price, size float64) {
	ob.mu.RLock()
	defer ob.mu.RUnlock()
	if len(ob.Asks) == 0 {
		return 0, 0
	}
	return ob.Asks[0].Price, ob.Asks[0].Size
}

// Spread returns the bid-ask spread.
func (ob *OrderBook) Spread() float64 {
	bp, _ := ob.BestBid()
	ap, _ := ob.BestAsk()
	if bp <= 0 || ap <= 0 {
		return 0
	}
	return ap - bp
}

// SpreadBps returns the spread in basis points.
func (ob *OrderBook) SpreadBps() float64 {
	bp, _ := ob.BestBid()
	ap, _ := ob.BestAsk()
	if bp <= 0 || ap <= 0 {
		return 0
	}
	mid := (bp + ap) / 2
	if mid == 0 {
		return 0
	}
	return (ap - bp) / mid * 10000
}

// MidPrice returns (bid+ask)/2.
func (ob *OrderBook) MidPrice() float64 {
	bp, _ := ob.BestBid()
	ap, _ := ob.BestAsk()
	if bp <= 0 || ap <= 0 {
		return 0
	}
	return (bp + ap) / 2
}

// VWAP computes the volume-weighted average price for the top n levels.
func (ob *OrderBook) VWAP(n int) (bidVWAP, askVWAP float64) {
	ob.mu.RLock()
	defer ob.mu.RUnlock()

	bidVWAP = vwap(ob.Bids, n)
	askVWAP = vwap(ob.Asks, n)
	return
}

// ToQuote converts the top-of-book to a Quote.
func (ob *OrderBook) ToQuote() Quote {
	ob.mu.RLock()
	defer ob.mu.RUnlock()

	var bp, bs, ap, as float64
	if len(ob.Bids) > 0 {
		bp = ob.Bids[0].Price
		bs = ob.Bids[0].Size
	}
	if len(ob.Asks) > 0 {
		ap = ob.Asks[0].Price
		as = ob.Asks[0].Size
	}
	return Quote{
		Symbol:    ob.Symbol,
		Timestamp: ob.AsOf,
		BidPrice:  bp,
		BidSize:   bs,
		AskPrice:  ap,
		AskSize:   as,
		Source:    "orderbook",
	}
}

// Depth returns up to n levels on each side.
func (ob *OrderBook) Depth(n int) (bids, asks []PriceLevel) {
	ob.mu.RLock()
	defer ob.mu.RUnlock()
	if n > len(ob.Bids) {
		n = len(ob.Bids)
	}
	bids = make([]PriceLevel, n)
	copy(bids, ob.Bids[:n])

	askN := n
	if askN > len(ob.Asks) {
		askN = len(ob.Asks)
	}
	asks = make([]PriceLevel, askN)
	copy(asks, ob.Asks[:askN])
	return
}

// Clear resets the order book.
func (ob *OrderBook) Clear() {
	ob.mu.Lock()
	ob.Bids = nil
	ob.Asks = nil
	ob.mu.Unlock()
}

// ImbalanceRatio returns (bid_vol - ask_vol) / (bid_vol + ask_vol) for top n levels.
// Ranges from -1 (pure ask pressure) to +1 (pure bid pressure).
func (ob *OrderBook) ImbalanceRatio(n int) float64 {
	ob.mu.RLock()
	defer ob.mu.RUnlock()

	bidVol := totalVol(ob.Bids, n)
	askVol := totalVol(ob.Asks, n)
	total := bidVol + askVol
	if total == 0 {
		return 0
	}
	return (bidVol - askVol) / total
}

// String returns a human-readable snapshot of the book.
func (ob *OrderBook) String() string {
	ob.mu.RLock()
	defer ob.mu.RUnlock()
	n := 5
	bids := ob.Bids
	asks := ob.Asks
	if len(bids) > n {
		bids = bids[:n]
	}
	if len(asks) > n {
		asks = asks[:n]
	}
	s := fmt.Sprintf("OrderBook[%s] spread=%.4f\n", ob.Symbol, ob.Spread())
	s += "  ASKS:\n"
	for i := len(asks) - 1; i >= 0; i-- {
		s += fmt.Sprintf("    %.4f x %.2f\n", asks[i].Price, asks[i].Size)
	}
	s += "  BIDS:\n"
	for _, l := range bids {
		s += fmt.Sprintf("    %.4f x %.2f\n", l.Price, l.Size)
	}
	return s
}

// ---- helpers ----

// updateLevel inserts or removes a price level in a sorted slice.
// ascending=true sorts ascending (for asks), false=descending (for bids).
func updateLevel(levels []PriceLevel, price, size float64, ascending bool) []PriceLevel {
	// Find existing level by price.
	for i, l := range levels {
		if math.Abs(l.Price-price) < 1e-12 {
			if size == 0 {
				return append(levels[:i], levels[i+1:]...)
			}
			levels[i].Size = size
			return levels
		}
	}
	if size == 0 {
		return levels
	}
	levels = append(levels, PriceLevel{Price: price, Size: size})
	if ascending {
		sort.Slice(levels, func(i, j int) bool { return levels[i].Price < levels[j].Price })
	} else {
		sort.Slice(levels, func(i, j int) bool { return levels[i].Price > levels[j].Price })
	}
	return levels
}

func vwap(levels []PriceLevel, n int) float64 {
	if n > len(levels) {
		n = len(levels)
	}
	var pv, v float64
	for i := 0; i < n; i++ {
		pv += levels[i].Price * levels[i].Size
		v += levels[i].Size
	}
	if v == 0 {
		return 0
	}
	return pv / v
}

func totalVol(levels []PriceLevel, n int) float64 {
	if n > len(levels) {
		n = len(levels)
	}
	var vol float64
	for i := 0; i < n; i++ {
		vol += levels[i].Size
	}
	return vol
}

// OrderBookRegistry manages OrderBooks for multiple symbols.
type OrderBookRegistry struct {
	mu    sync.RWMutex
	books map[string]*OrderBook
}

// NewOrderBookRegistry creates an OrderBookRegistry.
func NewOrderBookRegistry() *OrderBookRegistry {
	return &OrderBookRegistry{books: make(map[string]*OrderBook)}
}

// GetOrCreate returns the OrderBook for symbol, creating it if needed.
func (r *OrderBookRegistry) GetOrCreate(symbol string) *OrderBook {
	r.mu.RLock()
	ob, ok := r.books[symbol]
	r.mu.RUnlock()
	if ok {
		return ob
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	ob, ok = r.books[symbol]
	if !ok {
		ob = NewOrderBook(symbol)
		r.books[symbol] = ob
	}
	return ob
}

// Get returns the OrderBook for symbol, or nil.
func (r *OrderBookRegistry) Get(symbol string) *OrderBook {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.books[symbol]
}

// AllQuotes returns top-of-book quotes for all symbols.
func (r *OrderBookRegistry) AllQuotes() []Quote {
	r.mu.RLock()
	syms := make([]string, 0, len(r.books))
	for sym := range r.books {
		syms = append(syms, sym)
	}
	r.mu.RUnlock()

	quotes := make([]Quote, 0, len(syms))
	for _, sym := range syms {
		ob := r.Get(sym)
		if ob != nil {
			q := ob.ToQuote()
			if q.BidPrice > 0 || q.AskPrice > 0 {
				quotes = append(quotes, q)
			}
		}
	}
	return quotes
}

File: execution/orderbook/orderbook.py
"""
execution/orderbook/orderbook.py
=================================
Pure-Python L2 order-book data structure.

Thread-safe; all public properties and methods acquire a lock so an async feed
writer and a synchronous strategy reader can coexist safely.

Design notes
------------
- Bids and asks are stored as ``dict[float, float]`` (price -> cumulative qty).
- ``best_bid`` / ``best_ask`` scan sorted keys — O(n) but n is always small
  (≤ 25 levels in practice) so this avoids the overhead of a heap.
- ``vwap_to_fill`` walks the book and raises ``InsufficientLiquidityError`` if
  the book does not have enough depth to fill the requested quantity.
"""

from __future__ import annotations

import math
import threading
from typing import Optional


class InsufficientLiquidityError(Exception):
    """Raised when the book cannot absorb the requested fill quantity."""


class OrderBook:
    """
    Level-2 order book for a single trading symbol.

    Parameters
    ----------
    symbol : str
        Instrument identifier (e.g. ``"BTC/USD"``).
    max_levels : int
        Maximum number of price levels to retain per side (default 25).
    """

    def __init__(self, symbol: str, max_levels: int = 25) -> None:
        self.symbol = symbol
        self._max_levels = max_levels
        self._bids: dict[float, float] = {}   # price -> qty
        self._asks: dict[float, float] = {}
        self._lock = threading.Lock()
        self._last_update_ts: float = 0.0      # epoch seconds

    # ------------------------------------------------------------------
    # Mutation
    # ------------------------------------------------------------------

    def update(self, side: str, price: float, qty: float) -> None:
        """
        Apply a single level update.

        Parameters
        ----------
        side : str
            ``"bid"`` or ``"ask"`` (case-insensitive).
        price : float
            Price level to update.
        qty : float
            New quantity at this level.  ``qty == 0`` removes the level.
        """
        import time as _time

        side = side.lower()
        if side not in ("bid", "ask"):
            raise ValueError(f"side must be 'bid' or 'ask', got {side!r}")

        with self._lock:
            book = self._bids if side == "bid" else self._asks
            if qty == 0.0:
                book.pop(price, None)
            else:
                book[price] = qty
            # Trim to max levels, keeping best prices
            if len(book) > self._max_levels:
                if side == "bid":
                    # Keep highest bids
                    for p in sorted(book.keys())[: -self._max_levels]:
                        del book[p]
                else:
                    # Keep lowest asks
                    for p in sorted(book.keys())[self._max_levels :]:
                        del book[p]
            self._last_update_ts = _time.time()

    def apply_snapshot(
        self,
        bids: list[tuple[float, float]],
        asks: list[tuple[float, float]],
    ) -> None:
        """
        Replace the entire book with a fresh snapshot.

        Parameters
        ----------
        bids : list of (price, qty)
        asks : list of (price, qty)
        """
        import time as _time

        with self._lock:
            self._bids = {p: q for p, q in bids if q > 0}
            self._asks = {p: q for p, q in asks if q > 0}
            self._last_update_ts = _time.time()

    # ------------------------------------------------------------------
    # Core properties
    # ------------------------------------------------------------------

    @property
    def best_bid(self) -> Optional[float]:
        """Highest bid price, or None if empty."""
        with self._lock:
            return max(self._bids.keys()) if self._bids else None

    @property
    def best_ask(self) -> Optional[float]:
        """Lowest ask price, or None if empty."""
        with self._lock:
            return min(self._asks.keys()) if self._asks else None

    @property
    def mid_price(self) -> Optional[float]:
        """Arithmetic midpoint of best bid and ask, or None."""
        with self._lock:
            if not self._bids or not self._asks:
                return None
            bb = max(self._bids.keys())
            ba = min(self._asks.keys())
            return (bb + ba) / 2.0

    @property
    def spread_bps(self) -> Optional[float]:
        """
        Bid-ask spread expressed in basis points.

        ``spread_bps = (ask - bid) / mid * 10_000``

        Returns None if either side is empty or mid is zero.
        """
        with self._lock:
            if not self._bids or not self._asks:
                return None
            bb = max(self._bids.keys())
            ba = min(self._asks.keys())
            mid = (bb + ba) / 2.0
            if mid == 0:
                return None
            return (ba - bb) / mid * 10_000.0

    @property
    def last_update_ts(self) -> float:
        """Unix timestamp of the most recent update."""
        return self._last_update_ts

    # ------------------------------------------------------------------
    # Depth snapshot
    # ------------------------------------------------------------------

    def depth(self, n: int = 5) -> dict[str, list[tuple[float, float]]]:
        """
        Return top-N levels for each side.

        Returns
        -------
        dict with keys ``"bids"`` and ``"asks"``, each a list of
        ``(price, qty)`` tuples ordered best-to-worst.
        """
        with self._lock:
            bids = sorted(self._bids.items(), reverse=True)[:n]
            asks = sorted(self._asks.items())[:n]
        return {"bids": bids, "asks": asks}

    # ------------------------------------------------------------------
    # Imbalance
    # ------------------------------------------------------------------

    @property
    def imbalance(self) -> Optional[float]:
        """
        Order-flow imbalance at the top-5 levels.

        ``imbalance = (bid_qty - ask_qty) / (bid_qty + ask_qty)``

        Range [-1, 1].  Positive values indicate bid pressure.
        Returns None when either side is empty.
        """
        with self._lock:
            bid_qty = sum(
                q for _, q in sorted(self._bids.items(), reverse=True)[:5]
            )
            ask_qty = sum(
                q for _, q in sorted(self._asks.items())[:5]
            )
        total = bid_qty + ask_qty
        if total == 0:
            return None
        return (bid_qty - ask_qty) / total

    # ------------------------------------------------------------------
    # VWAP fill estimator
    # ------------------------------------------------------------------

    def vwap_to_fill(self, side: str, qty: float) -> float:
        """
        Walk the book and compute the volume-weighted average fill price
        for a given quantity.

        Parameters
        ----------
        side : str
            ``"buy"`` — walks the ask side (ascending price).
            ``"sell"`` — walks the bid side (descending price).
        qty : float
            Quantity to fill.

        Returns
        -------
        float
            Expected VWAP fill price.

        Raises
        ------
        InsufficientLiquidityError
            If available depth is smaller than ``qty``.
        ValueError
            If ``side`` is invalid or ``qty`` <= 0.
        """
        if qty <= 0:
            raise ValueError("qty must be positive")
        side = side.lower()
        if side not in ("buy", "sell"):
            raise ValueError(f"side must be 'buy' or 'sell', got {side!r}")

        with self._lock:
            if side == "buy":
                levels = sorted(self._asks.items())          # ascending ask
            else:
                levels = sorted(self._bids.items(), reverse=True)  # descending bid

        remaining = qty
        notional = 0.0
        for price, level_qty in levels:
            fill = min(remaining, level_qty)
            notional += fill * price
            remaining -= fill
            if remaining <= 1e-12:
                break

        if remaining > 1e-12:
            raise InsufficientLiquidityError(
                f"Book for {self.symbol} has insufficient {side} liquidity: "
                f"need {qty}, only {qty - remaining:.6f} available"
            )

        return notional / qty

    # ------------------------------------------------------------------
    # Convenience
    # ------------------------------------------------------------------

    def bid_depth_usd(self, n: int = 5) -> float:
        """Total USD value of the top-N bid levels."""
        with self._lock:
            return sum(
                p * q for p, q in sorted(self._bids.items(), reverse=True)[:n]
            )

    def ask_depth_usd(self, n: int = 5) -> float:
        """Total USD value of the top-N ask levels."""
        with self._lock:
            return sum(
                p * q for p, q in sorted(self._asks.items())[:n]
            )

    def __repr__(self) -> str:
        bb = self.best_bid
        ba = self.best_ask
        sp = self.spread_bps
        return (
            f"<OrderBook {self.symbol} "
            f"bid={bb:.4f} ask={ba:.4f} spread={sp:.2f}bps>"
            if bb and ba and sp
            else f"<OrderBook {self.symbol} empty>"
        )
